lib/remote/actionshandler.cpp

```diff
@@ -56,9 +56,14 @@ bool ActionsHandler::HandleRequest(
 
 		try {
 			objs = FilterUtility::GetFilterTargets(qd, params, user);
-		} catch (const std::exception& ex) {
-			HttpUtility::SendJsonError(response, params, 404,
-				"No objects found.",
+		} catch (const std::invalid_argument& ex) {
+			HttpUtility::SendJsonError(response, params, 400,
+				"Invalid filter.",
+				DiagnosticInformation(ex));
+			return true;
+		} catch (const ScriptError& ex) {
+			HttpUtility::SendJsonError(response, params, 401,
+				"Access denied for filter.",
 				DiagnosticInformation(ex));
 			return true;
 		}
```

Comment:
403 would probably be more appropriate than 401 in the situation you described. But anyways, `ScriptError` is much more generic, so I don't think it's valid to conclude access denied from that exception type.

Reply:
I concluded by checking reachable throws and it happens that all `ScriptError`s seem to indicate kind of "Access denied". But agreed, that's brittle. Maybe introduce a new exception type `PermissionError`?  Or add an enum member `m_Reason` to `ScriptError`?

Otherwise we just can't indicate `403`. It'll remains hidden behind a more generic error code.

Reply:
> reachable throws and it happens that all `ScriptError`s seem to indicate kind of "Access denied"

Which ones?

Reply:
> Which ones?

[Registered API actions](https://github.com/Icinga/icinga2/blob/v2.13.2/lib/icinga/apiactions.cpp#L26) use `QueryDescription.Types` `"Service"`, `"Host"` and [enter](https://github.com/Icinga/icinga2/blob/v2.13.2/lib/remote/actionshandler.cpp#L58) throwing code at [FilterUtility::GetFilterTargets](https://github.com/Icinga/icinga2/blob/v2.13.2/lib/remote/filterutility.cpp#L180), where they can directly hit
- [`ScriptError("Missing permission: " + requiredPermission)` in filterutility.cpp#L176](https://github.com/Icinga/icinga2/blob/v2.13.2/lib/remote/filterutility.cpp#L176)
- [`ScriptError("Access denied to object '" + name + "' of type '" + type + "'")` in filterutility.cpp#L225](https://github.com/Icinga/icinga2/blob/v2.13.2/lib/remote/filterutility.cpp#L225)

I see we can hit other kinds of `ScriptError` farther down the stack, e.g. from [expression.cpp](https://github.com/Icinga/icinga2/blob/v2.13.2/lib/config/expression.cpp). So handling `ScriptError` unconditionally as permission problem is definitely wrong.